File: python/examples/demo_progress_and_roots.py
"""
Interactive Live Demo: Progress Streaming & Reverse Roots Discovery.

Demonstrates:
1. [Server -> Client] Roots Discovery (`roots/list`): Server queries Client's workspace folders.
2. [Server -> Client] Progress Streaming (`notifications/progress`): Server pushes real-time progress bars.
3. [Server -> Client] Live Logging (`notifications/message`): Server streams log notifications unprompted.
"""

import asyncio
import sys
import websockets
from mcp_websocket import (
    JSONRPCNotification,
    JSONRPCRequest,
    JSONRPCResponse,
    WebSocketClientTransport,
    serve_websocket,
)

if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")


class DemoServer:
    def __init__(self):
        self.client_roots = []
        self.pending_client_requests = {}
        self.write_stream = None

    async def run(self, read_stream, write_stream):
        self.write_stream = write_stream
        async with asyncio.TaskGroup() as tg:
            async for msg in read_stream:
                if not isinstance(msg, dict):
                    continue

                msg_id = str(msg.get("id"))
                # If this is a response to a server-initiated request (roots/list)
                if msg_id in self.pending_client_requests and "result" in msg:
                    fut = self.pending_client_requests.pop(msg_id)
                    fut.set_result(msg["result"])
                    continue

                method = msg.get("method")
                req_id = msg.get("id")

                # Handshake
                if method == "initialize":
                    await write_stream.send(
                        JSONRPCResponse(
                            id=req_id,
                            result={
                                "protocolVersion": "2024-11-05",
                                "capabilities": {"tools": {}, "logging": {}},
                                "serverInfo": {"name": "progress-and-roots-demo", "version": "0.1.0"},
                            },
                        )
                    )

                    # SERVER -> CLIENT: Reverse query for Client Roots!
                    async def fetch_roots():
                        await asyncio.sleep(0.05)
                        print("   ⚡ [Server] Initiating Server -> Client reverse request: `roots/list`...")
                        fut = asyncio.get_running_loop().create_future()
                        self.pending_client_requests["srv-roots-1"] = fut
                        await write_stream.send(
                            JSONRPCRequest(id="srv-roots-1", method="roots/list", params={})
                        )
                        roots_res = await fut
                        self.client_roots = roots_res.get("roots", [])
                        print(f"   📂 [Server] Successfully discovered Client roots: {self.client_roots}\n")

                    tg.create_task(fetch_roots())

                # Tool call (triggers progress streaming)
                elif method == "tools/call":
                    params = msg.get("params", {})
                    tool_name = params.get("name")

                    if tool_name == "stream_download_task":
                        total_steps = 5
                        token = params.get("_meta", {}).get("progressToken", "dl-token-1")

                        # Send Log Notification
                        await write_stream.send(
                            JSONRPCNotification(
                                method="notifications/message",
                                params={"level": "info", "logger": "downloader", "data": "Starting 5-step data download..."},
                            )
                        )

                        # Stream Progress frames
                        for step in range(1, total_steps + 1):
                            await asyncio.sleep(0.4)  # Simulate work
                            await write_stream.send(
                                JSONRPCNotification(
                                    method="notifications/progress",
                                    params={"progressToken": token, "progress": step, "total": total_steps},
                                )
                            )

                        await write_stream.send(
                            JSONRPCResponse(
                                id=req_id,
                                result={"content": [{"type": "text", "text": "Download and indexing completed 100%!"}]},
                            )
                        )


def render_progress_bar(progress: int, total: int, width: int = 25):
    """Renders a live terminal progress bar."""
    percent = int((progress / total) * 100)
    filled = int(width * progress // total)
    bar = "█" * filled + "░" * (width - filled)
    sys.stdout.write(f"\r   📊 [Client Live Stream] [{bar}] {percent}% (Step {progress}/{total})")
    sys.stdout.flush()
    if progress == total:
        sys.stdout.write("\n")


async def run_demo():
    print("=" * 70)
    print("🚀 Live Demo: Progress Streaming & Reverse Roots Discovery over WebSocket")
    print("=" * 70)

    server = DemoServer()

    async def ws_handler(ws):
        await serve_websocket(server, ws)

    async with websockets.serve(ws_handler, "localhost", 8769):
        url = "ws://localhost:8769"

        async with WebSocketClientTransport(url) as (read_stream, write_stream):
            # 1. Client sends Handshake (declaring roots capability)
            print("\n1️⃣ [Handshake] Client connects and declares roots capability...")
            await write_stream.send(
                JSONRPCRequest(
                    id=1,
                    method="initialize",
                    params={
                        "protocolVersion": "2024-11-05",
                        "capabilities": {"roots": {"listChanged": True}},
                        "clientInfo": {"name": "demo-client", "version": "0.1.0"},
                    },
                )
            )

            # Receive initialize response
            init_res = await read_stream.receive()
            print("   ✅ Handshake response received.")

            # 2. Receive Server-Initiated `roots/list` request!
            roots_req = await read_stream.receive()
            if roots_req.get("method") == "roots/list":
                print(f"   📥 [Client] Received Server-Initiated request: `roots/list` (id: {roots_req.get('id')})")
                print("   📤 [Client] Responding with open project directories...")
                await write_stream.send(
                    JSONRPCResponse(
                        id=roots_req.get("id"),
                        result={
                            "roots": [
                                {"uri": "file:///i:/01-Master_Code/Apps/MCP-WebSocket-Transport", "name": "MCP-WebSocket-Transport"},
                                {"uri": "file:///workspace/ai-agent", "name": "AI Agent Workspace"},
                            ]
                        },
                    )
                )

            await asyncio.sleep(0.2)

            # 3. Client calls heavy task and requests progress streaming
            print("\n2️⃣ [Client -> Server] Invoking tool `stream_download_task`...")
            await write_stream.send(
                JSONRPCRequest(
                    id=2,
                    method="tools/call",
                    params={
                        "name": "stream_download_task",
                        "_meta": {"progressToken": "download-job-42"},
                    },
                )
            )

            print("\n3️⃣ [Server -> Client] Streaming live progress & logs unprompted:")
            while True:
                msg = await read_stream.receive()
                if not isinstance(msg, dict):
                    continue

                method = msg.get("method")

                # Live Log Notification
                if method == "notifications/message":
                    print(f"   📢 [Log Notification]: {msg['params']['data']}")

                # Live Progress Notification
                elif method == "notifications/progress":
                    p = msg["params"]["progress"]
                    tot = msg["params"]["total"]
                    render_progress_bar(p, tot)

                # Final Tool Response
                elif msg.get("id") == 2:
                    print(f"\n4️⃣ [Result] Tool Completed: {msg['result']['content'][0]['text']}")
                    break

    print("\n🎉 Verification of Progress Streaming & Reverse Roots Discovery 100% Successful!")


if __name__ == "__main__":
    asyncio.run(run_demo())

File: typescript/examples/demo-progress-and-roots.ts
import { WebSocket, WebSocketServer } from "ws";
import {
  WebSocketClientTransport,
  WebSocketServerTransport,
} from "../src/index.js";

/**
 * TypeScript Live Demo: Progress Streaming & Reverse Roots Discovery
 */
async function runDemo() {
  console.log("=".repeat(70));
  console.log(
    "🚀 TypeScript Demo: Progress Streaming & Reverse Roots Discovery",
  );
  console.log("=".repeat(70));

  const port = 8770;
  const wss = new WebSocketServer({ port });

  const pendingRequests = new Map();
  let clientRoots = [];

  wss.on("connection", async (ws) => {
    const serverTransport = new WebSocketServerTransport(ws);

    serverTransport.onmessage = async (msg: any) => {
      // Check if response to server-initiated request
      if (
        msg.id &&
        pendingRequests.has(msg.id) &&
        (msg.result !== undefined || msg.error !== undefined)
      ) {
        const resolve = pendingRequests.get(msg.id);
        pendingRequests.delete(msg.id);
        resolve(msg);
        return;
      }

      if (msg.method === "initialize") {
        await serverTransport.send({
          jsonrpc: "2.0",
          id: msg.id,
          result: {
            protocolVersion: "2024-11-05",
            capabilities: { tools: {}, logging: {} },
            serverInfo: { name: "ts-demo-server", version: "0.1.0" },
          },
        });

        // 1. SERVER -> CLIENT: Reverse query for Client Roots!
        setTimeout(async () => {
          console.log(
            "   ⚡ [Server] Initiating Server -> Client reverse request: `roots/list`...",
          );
          const rootsPromise = new Promise((res) =>
            pendingRequests.set("srv-roots-1", res),
          );
          await serverTransport.send({
            jsonrpc: "2.0",
            id: "srv-roots-1",
            method: "roots/list",
            params: {},
          });
          const rootsRes: any = await rootsPromise;
          clientRoots = rootsRes.result?.roots || [];
          console.log(
            `   📂 [Server] Successfully discovered Client roots:`,
            clientRoots,
            "\n",
          );
        }, 50);
      } else if (msg.method === "tools/call") {
        if (msg.params?.name === "heavy_computation_task") {
          const token = msg.params._meta?.progressToken || "job-token-1";
          const totalSteps = 5;

          // Send Log Notification
          await serverTransport.send({
            jsonrpc: "2.0",
            method: "notifications/message",
            params: {
              level: "info",
              logger: "computation",
              data: "Starting 5-step distributed task...",
            },
          });

          // Stream Progress frames unprompted
          for (let step = 1; step <= totalSteps; step++) {
            await new Promise((r) => setTimeout(r, 300));
            await serverTransport.send({
              jsonrpc: "2.0",
              method: "notifications/progress",
              params: {
                progressToken: token,
                progress: step,
                total: totalSteps,
              },
            });
          }

          await serverTransport.send({
            jsonrpc: "2.0",
            id: msg.id,
            result: {
              content: [
                {
                  type: "text",
                  text: "Heavy computation task completed with 100% success.",
                },
              ],
            },
          });
        }
      }
    };

    await serverTransport.start();
  });

  // Client side
  const clientTransport = new WebSocketClientTransport(
    `ws://localhost:${port}`,
    { WebSocket },
  );

  const pendingClientResolutions = new Map();

  function renderProgressBar(
    progress: number,
    total: number,
    width: number = 25,
  ) {
    const percent = Math.floor((progress / total) * 100);
    const filled = Math.floor((width * progress) / total);
    const bar = "█".repeat(filled) + "░".repeat(width - filled);
    process.stdout.write(
      `\r   📊 [Client Live Stream] [${bar}] ${percent}% (Step ${progress}/${total})`,
    );
    if (progress === total) {
      process.stdout.write("\n");
    }
  }

  clientTransport.onmessage = async (msg: any) => {
    // Check for Server -> Client roots request
    if (msg.method === "roots/list") {
      console.log(
        `   📥 [Client] Received Server-Initiated request: \`roots/list\` (id: ${msg.id})`,
      );
      console.log("   📤 [Client] Responding with open project directories...");
      await clientTransport.send({
        jsonrpc: "2.0",
        id: msg.id,
        result: {
          roots: [
            {
              uri: "file:///i:/01-Master_Code/Apps/MCP-WebSocket-Transport",
              name: "MCP-WebSocket-Transport",
            },
            {
              uri: "file:///workspace/node-app",
              name: "Node App Workspace",
            },
          ],
        },
      });
      return;
    }

    // Check for log notification
    if (msg.method === "notifications/message") {
      console.log(`   📢 [Log Notification]: ${msg.params.data}`);
      return;
    }

    // Check for progress notification
    if (msg.method === "notifications/progress") {
      renderProgressBar(msg.params.progress, msg.params.total);
      return;
    }

    if (msg.id && pendingClientResolutions.has(msg.id)) {
      const resolve = pendingClientResolutions.get(msg.id);
      pendingClientResolutions.delete(msg.id);
      resolve(msg);
    }
  };

  await clientTransport.start();

  // 1. Handshake
  console.log(
    "\n1️⃣ [Handshake] Client connects and declares roots capability...",
  );
  await new Promise((res) => {
    pendingClientResolutions.set(1, res);
    clientTransport.send({
      jsonrpc: "2.0",
      id: 1,
      method: "initialize",
      params: {
        protocolVersion: "2024-11-05",
        capabilities: { roots: { listChanged: true } },
        clientInfo: { name: "demo-ts-client", version: "0.1.0" },
      },
    });
  });
  console.log("   ✅ Handshake response received.");

  await new Promise((r) => setTimeout(r, 200));

  // 2. Call Tool
  console.log(
    "\n2️⃣ [Client -> Server] Invoking tool `heavy_computation_task`...",
  );
  console.log(
    "\n3️⃣ [Server -> Client] Streaming live progress & logs unprompted:",
  );
  const toolResult: any = await new Promise((res) => {
    pendingClientResolutions.set(2, res);
    clientTransport.send({
      jsonrpc: "2.0",
      id: 2,
      method: "tools/call",
      params: {
        name: "heavy_computation_task",
        _meta: { progressToken: "job-token-1" },
      },
    });
  });

  console.log(
    `\n4️⃣ [Result] Tool Completed: ${toolResult.result.content[0].text}`,
  );

  await clientTransport.close();
  wss.close();
  console.log(
    "\n🎉 TypeScript Progress Streaming & Roots Discovery 100% Successful!",
  );
}

runDemo().catch(console.error);
